### ISSUE TITLE
Remove client id and secret from the form for custom identity providers

### ISSUE THREAD

**Issue by ullgren** [keycloak/keycloak#21891]
### Description

The current implementation to support custom identity providers in #15344 adds client id and secret as mandatory fields.

This is a proposal to remove these for implementations that are not "Social Identity Providers".

### Motivation

Not all, perhaps most, of the custom identity providers does not have the need or concept of client id or secret.
As an example the [BankID provider](https://github.com/sweid4keycloak/bankid4keycloak) or [CAS Identity Provider](https://github.com/RoboJackets/keycloak-cas/).

Also the way adding a custom identity provider prior to version 21.x allowed you to add a form with out the client id and secret.

### Details

A simple implementation would be to change `js/apps/admin-ui/src/identity-providers/add/GeneralSettings.tsx` so that it only adds the client id and secret fields if the provider implements `org.keycloak.broker.social.SocialIdentityProvider` 

@evan10s @edewit Any input on this ?

**Comment by ullgren** [comment]
[Here](https://github.com/keycloak/keycloak/compare/main...Redpill-Linpro:keycloak:21891-remove-client-id-secret-for-non-social-providers) are a proposed implementation. Please let me know if you are interested in a PR

**Comment by jonkoops** [comment]
If we are going to filter out these fields, I think we should do so on the server-side and only render what the API gives back on the client.

**Comment by jonkoops** [comment]
> But, I gotta admit, I'm a bit concerned that this might take more time to implement since it would require a bigger change.

That is alright with me, these kinds of changes have a habit of becoming permanent, and I'd rather solve it properly.

**Comment by sschu** [comment]
I am not sure the existence of client id/secret is directly related to the fact that an IDP is a social IDP - it really depends on the protocol being implemented and any IDP doing something based on OIDC will probably have client id/secret.

**Comment by ullgren** [comment]
> I am not sure the existence of client id/secret is directly related to the fact that an IDP is a social IDP - it really depends on the protocol being implemented and any IDP doing something based on OIDC will probably have client id/secret.

@sschu  Again agree. It is just that in the current implementation it is determined by the [provider id string](https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L302-L303) if the [OIDC, SAML or just a generic "GeneralSettings"](https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L339-L348) are displayed.  The "GeneralSettings" are therefor used by all out of the box Social IDPs and any custom providers. And it includes the Client ID/Secret. 

This seems to indicate that the assumption is that any social, or custom, provider uses OIDC.

**Comment by Nowheresly** [comment]
I agree with @ullgren approach.

**Comment by Nowheresly** [comment]
I agree with @ullgren approach. The current implementation of the admin UI relies on the provider id string to guess if it's a saml or an oidc identity provider. So the implementation of the PR seems consistent with what have been done so far...
I agree as well with the comment of @jonkoops :

> If we are going to filter out these fields, I think we should do so on the server-side and only render what the API gives back on the client.

but then, IMHO, we should consider doing this refactoring as another PR since it is a different topic.

The proposed PR helps fixing the way we can create a custom identity provider. 
Is there any chances to see this PR merged or at least further considered?

**Comment by jonkoops** [comment]
You can always open up a PR and we can discuss the implementation work and if it's up to snuff. If you want to get some quick feedback on an early implementation the PR can be marked as draft.

**Comment by ullgren** [comment]
@jonkoops @Nowheresly I will re-base my currently proposed changes and open a draft PR on it.
I'm also into a better solution where this decision is made on the server side by the provider but this work takes a bit longer than expected.
